
<!DOCTYPE html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      <meta http-equiv="x-ua-compatible" content="ie=edge">
      
        <meta name="description" content="A julia module to perform trait simulation">
      
      
      
        <meta name="author" content="Huwenbo Shi">
      
      
        <link rel="shortcut icon" href="../../assets/images/favicon.png">
      
      <meta name="generator" content="mkdocs-0.16.3, mkdocs-material-1.8.1">
    
    
      
        <title>Examples - TraitSimulation.jl</title>
      
    
    
      <script src="../../assets/javascripts/modernizr-1df76c4e58.js"></script>
    
    
      <link rel="stylesheet" href="../../assets/stylesheets/application-bfecc7305d.css">
      
        <link rel="stylesheet" href="../../assets/stylesheets/application-02c2a4388f.palette.css">
      
    
    
      
        
        
        
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,400,400i,700|Roboto+Mono">
        <style>body,input{font-family:"Roboto","Helvetica Neue",Helvetica,Arial,sans-serif}code,kbd,pre{font-family:"Roboto Mono","Courier New",Courier,monospace}</style>
      
      <link rel="stylesheet" href="https://fonts.googleapis.com/icon?family=Material+Icons">
    
    
      <link rel="stylesheet" href="../../assets/Documenter.css">
    
    
  </head>
  
  
  
  
    <body data-md-color-primary="indigo" data-md-color-accent="blue">
  
    <svg class="md-svg">
      <defs>
        
        
          <svg xmlns="http://www.w3.org/2000/svg" width="416" height="448" viewBox="0 0 416 448" id="github"><path fill="currentColor" d="M160 304q0 10-3.125 20.5t-10.75 19T128 352t-18.125-8.5-10.75-19T96 304t3.125-20.5 10.75-19T128 256t18.125 8.5 10.75 19T160 304zm160 0q0 10-3.125 20.5t-10.75 19T288 352t-18.125-8.5-10.75-19T256 304t3.125-20.5 10.75-19T288 256t18.125 8.5 10.75 19T320 304zm40 0q0-30-17.25-51T296 232q-10.25 0-48.75 5.25Q229.5 240 208 240t-39.25-2.75Q130.75 232 120 232q-29.5 0-46.75 21T56 304q0 22 8 38.375t20.25 25.75 30.5 15 35 7.375 37.25 1.75h42q20.5 0 37.25-1.75t35-7.375 30.5-15 20.25-25.75T360 304zm56-44q0 51.75-15.25 82.75-9.5 19.25-26.375 33.25t-35.25 21.5-42.5 11.875-42.875 5.5T212 416q-19.5 0-35.5-.75t-36.875-3.125-38.125-7.5-34.25-12.875T37 371.5t-21.5-28.75Q0 312 0 260q0-59.25 34-99-6.75-20.5-6.75-42.5 0-29 12.75-54.5 27 0 47.5 9.875t47.25 30.875Q171.5 96 212 96q37 0 70 8 26.25-20.5 46.75-30.25T376 64q12.75 25.5 12.75 54.5 0 21.75-6.75 42 34 40 34 99.5z"/></svg>
        
      </defs>
    </svg>
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="drawer">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="search">
    <label class="md-overlay" data-md-component="overlay" for="drawer"></label>
    
      <header class="md-header" data-md-component="header">
  <nav class="md-header-nav md-grid">
    <div class="md-flex">
      <div class="md-flex__cell md-flex__cell--shrink">
        
          <a href="../.." title="TraitSimulation.jl" class="md-icon md-icon--home md-header-nav__button">
          </a>
        
      </div>
      <div class="md-flex__cell md-flex__cell--shrink">
        <label class="md-icon md-icon--menu md-header-nav__button" for="drawer"></label>
      </div>
      <div class="md-flex__cell md-flex__cell--stretch">
        <span class="md-flex__ellipsis md-header-nav__title">
          
            
              
            
            Examples
          
        </span>
      </div>
      <div class="md-flex__cell md-flex__cell--shrink">
        
          <label class="md-icon md-icon--search md-header-nav__button" for="search"></label>
          
<div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="search"></label>
  <div class="md-search__inner">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" required placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="query">
      <label class="md-icon md-search__icon" for="search"></label>
      <button type="reset" class="md-icon md-search__icon" data-md-component="reset">close</button>
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" data-md-scrollfix>
        <div class="md-search-result" data-md-component="result" data-md-lang-search="" data-md-lang-tokenizer="[\s\-]+">
          <div class="md-search-result__meta" data-md-lang-result-none="No matching documents" data-md-lang-result-one="1 matching document" data-md-lang-result-other="# matching documents">
            Type to start searching
          </div>
          <ol class="md-search-result__list"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
        
      </div>
      
        <div class="md-flex__cell md-flex__cell--shrink">
          <div class="md-header-nav__source">
            


  


  <a href="https://github.com/huwenboshi/TraitSimulation.jl" title="Go to repository" class="md-source" data-md-source="github">
    
      <div class="md-source__icon">
        <svg viewBox="0 0 24 24" width="24" height="24">
          <use xlink:href="#github" width="24" height="24"></use>
        </svg>
      </div>
    
    <div class="md-source__repository">
      GitHub
    </div>
  </a>

          </div>
        </div>
      
    </div>
  </nav>
</header>
    
    <div class="md-container">
      
      
      <main class="md-main">
        <div class="md-main__inner md-grid" data-md-component="container">
          
            
              <div class="md-sidebar md-sidebar--primary" data-md-component="navigation">
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    <nav class="md-nav md-nav--primary" data-md-level="0">
  <label class="md-nav__title md-nav__title--site" for="drawer">
    
      <i class="md-icon md-icon--home md-nav__button"></i>
    
    TraitSimulation.jl
  </label>
  
    <div class="md-nav__source">
      


  


  <a href="https://github.com/huwenboshi/TraitSimulation.jl" title="Go to repository" class="md-source" data-md-source="github">
    
      <div class="md-source__icon">
        <svg viewBox="0 0 24 24" width="24" height="24">
          <use xlink:href="#github" width="24" height="24"></use>
        </svg>
      </div>
    
    <div class="md-source__repository">
      GitHub
    </div>
  </a>

    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
      


  <li class="md-nav__item">
    <a href="../.." title="Introduction" class="md-nav__link">
      Introduction
    </a>
  </li>

    
      
      
      

  


  <li class="md-nav__item md-nav__item--active">
    
    <input class="md-toggle md-nav__toggle" data-md-toggle="toc" type="checkbox" id="toc">
    
      
    
    
      <label class="md-nav__link md-nav__link--active" for="toc">
        Examples
      </label>
    
    <a href="./" title="Examples" class="md-nav__link md-nav__link--active">
      Examples
    </a>
    
      
<nav class="md-nav md-nav--secondary">
  
  
    
  
  
    <label class="md-nav__title" for="toc">Table of contents</label>
    <ul class="md-nav__list" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#overview" title="Overview" class="md-nav__link">
    Overview
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#simulate-a-single-trait" title="Simulate a single trait" class="md-nav__link">
    Simulate a single trait
  </a>
  
    <nav class="md-nav">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#simulate-a-trait-using-genotype-data-in-plink-format" title="Simulate a trait using genotype data in PLINK format" class="md-nav__link">
    Simulate a trait using genotype data in PLINK format
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#simulate-a-trait-with-pcs-as-covariates" title="Simulate a trait with PC's as covariates" class="md-nav__link">
    Simulate a trait with PC's as covariates
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#simulate-a-trait-under-random-effect-model" title="Simulate a trait under random effect model" class="md-nav__link">
    Simulate a trait under random effect model
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#simulate-a-trait-under-mixed-effect-model" title="Simulate a trait under mixed effect model" class="md-nav__link">
    Simulate a trait under mixed effect model
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#simulate-a-trait-with-non-normal-distributions" title="Simulate a trait with non-normal distributions" class="md-nav__link">
    Simulate a trait with non-normal distributions
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#simulate-multiple-traits" title="Simulate multiple traits" class="md-nav__link">
    Simulate multiple traits
  </a>
  
    <nav class="md-nav">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#simulate-three-independent-traits-with-different-response-distributions" title="Simulate three independent traits with different response distributions" class="md-nav__link">
    Simulate three independent traits with different response distributions
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#simulate-two-correlated-traits" title="Simulate two correlated traits" class="md-nav__link">
    Simulate two correlated traits
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
      
      
    </ul>
  
</nav>
    
  </li>

    
      
      
      


  <li class="md-nav__item">
    <a href="../../model/model/" title="Model Specifications" class="md-nav__link">
      Model Specifications
    </a>
  </li>

    
      
      
      


  <li class="md-nav__item">
    <a href="../../generate/generate/" title="Generate Simulations" class="md-nav__link">
      Generate Simulations
    </a>
  </li>

    
      
      
      


  <li class="md-nav__item">
    <a href="../../link/link/" title="Link Functions" class="md-nav__link">
      Link Functions
    </a>
  </li>

    
      
      
      


  <li class="md-nav__item">
    <a href="../../dist/dist/" title="Response Distributions" class="md-nav__link">
      Response Distributions
    </a>
  </li>

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              <div class="md-sidebar md-sidebar--secondary" data-md-component="toc">
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    
<nav class="md-nav md-nav--secondary">
  
  
    
  
  
    <label class="md-nav__title" for="toc">Table of contents</label>
    <ul class="md-nav__list" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#overview" title="Overview" class="md-nav__link">
    Overview
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#simulate-a-single-trait" title="Simulate a single trait" class="md-nav__link">
    Simulate a single trait
  </a>
  
    <nav class="md-nav">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#simulate-a-trait-using-genotype-data-in-plink-format" title="Simulate a trait using genotype data in PLINK format" class="md-nav__link">
    Simulate a trait using genotype data in PLINK format
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#simulate-a-trait-with-pcs-as-covariates" title="Simulate a trait with PC's as covariates" class="md-nav__link">
    Simulate a trait with PC's as covariates
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#simulate-a-trait-under-random-effect-model" title="Simulate a trait under random effect model" class="md-nav__link">
    Simulate a trait under random effect model
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#simulate-a-trait-under-mixed-effect-model" title="Simulate a trait under mixed effect model" class="md-nav__link">
    Simulate a trait under mixed effect model
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#simulate-a-trait-with-non-normal-distributions" title="Simulate a trait with non-normal distributions" class="md-nav__link">
    Simulate a trait with non-normal distributions
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#simulate-multiple-traits" title="Simulate multiple traits" class="md-nav__link">
    Simulate multiple traits
  </a>
  
    <nav class="md-nav">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#simulate-three-independent-traits-with-different-response-distributions" title="Simulate three independent traits with different response distributions" class="md-nav__link">
    Simulate three independent traits with different response distributions
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#simulate-two-correlated-traits" title="Simulate two correlated traits" class="md-nav__link">
    Simulate two correlated traits
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
      
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          <div class="md-content">
            <article class="md-content__inner md-typeset">
              
                
                  <a href="https://github.com/huwenboshi/TraitSimulation.jl/edit/master/docs/example/example.md" title="Edit this page" class="md-icon md-content__icon">edit</a>
                
                
                <p><a id='Getting-Started-1'></a></p>
<h1 id="getting-started">Getting Started</h1>
<p>This section provides some examples on how to use the TraitSimulation module.</p>
<p><a id='Overview-1'></a></p>
<h2 id="overview">Overview</h2>
<p>The TraitSimulation module allows users to simulate phenotypes under various models, including fixed-effect, random-effect, and mixed-effect models. The simulation procedure takes a two-step approach. In the first step, the user constructs a simulation model using the <code>SimulationModel</code> type. In the second step, the user calls the <code>simulate</code> function along with the data to generate the simulated phenotype, which is appended as a column to the input data frame.</p>
<p><a id='Simulate-a-single-trait-1'></a></p>
<h2 id="simulate-a-single-trait">Simulate a single trait</h2>
<p><a id='Simulate-a-trait-using-genotype-data-in-PLINK-format-1'></a></p>
<h3 id="simulate-a-trait-using-genotype-data-in-plink-format">Simulate a trait using genotype data in PLINK format</h3>
<p>TraitSimulation module can take genotype data in PLINK format through <a href="https://github.com/OpenMendel/SnpArrays.jl">SnpArrays</a>. For example, the following code reads in genotype data from the PLINK files &ldquo;hapmap3.bed&rdquo;, &ldquo;hapmap3.bim&rdquo;, &ldquo;hapmap3.fam&rdquo;, and then simulate the trait based on the first 3 SNPs using the <code>simulate</code>(see below for more details). More specifically, the code below samples phenotypes from</p>
<p>
<script type="math/tex; mode=display">
y \sim N(\mu, 1.0)
</script>
</p>
<p>where</p>
<p>
<script type="math/tex; mode=display">
\mu = x_1+0.1x_2x_3.
</script>
</p>
<div class="codehilite"><pre><span></span><span class="k">using</span> <span class="n">DataFrames</span><span class="p">,</span> <span class="n">Distributions</span><span class="p">,</span> <span class="n">SnpArrays</span><span class="p">,</span> <span class="n">TraitSimulation</span>

<span class="c"># Load genotype data in PLINK format</span>
<span class="n">data</span> <span class="o">=</span> <span class="n">SnpArray</span><span class="p">(</span><span class="s">&quot;hapmap3&quot;</span><span class="p">)</span>

<span class="c"># Create the fixed-effect simulation model</span>
<span class="n">sim_model</span> <span class="o">=</span> <span class="n">FixedEffectModel</span><span class="p">(</span><span class="nd">@formula</span><span class="p">(</span><span class="n">T</span> <span class="o">~</span> <span class="n">x1</span><span class="o">+</span><span class="mi">2</span><span class="n">x2</span><span class="o">*</span><span class="n">x3</span><span class="p">),</span>
                             <span class="n">IdentityLink</span><span class="p">(),</span> <span class="n">NormalResponse</span><span class="p">(</span><span class="mf">1.0</span><span class="p">))</span>

<span class="c"># Generate the simulation</span>
<span class="n">y</span> <span class="o">=</span> <span class="n">simulate</span><span class="p">(</span><span class="n">sim_model</span><span class="p">,</span> <span class="n">data</span><span class="p">)</span>
</pre></div>


<p><a id='Simulate-a-trait-with-PC's-as-covariates-1'></a></p>
<h3 id="simulate-a-trait-with-pcs-as-covariates">Simulate a trait with PC&rsquo;s as covariates</h3>
<p>To include principal components as covariates in the simulation procedure, one can apply <code>pca</code> in the <code>SnpArrays</code> module to obtain the PCs and then include them as a column in the input data frame. The following code simulate a trait using the first 3 SNPs and 2 PCs. More specifically, the code below samples phenotypes from</p>
<p>
<script type="math/tex; mode=display">
y \sim N(\mu, 1.0)
</script>
</p>
<p>where</p>
<p>
<script type="math/tex; mode=display">
\mu = x_1+0.1x_2x_3 + PC1 + PC2.
</script>
</p>
<div class="codehilite"><pre><span></span><span class="k">using</span> <span class="n">DataFrames</span><span class="p">,</span> <span class="n">Distributions</span><span class="p">,</span> <span class="n">SnpArrays</span><span class="p">,</span> <span class="n">TraitSimulation</span>

<span class="c"># Load Genotype data in PLINK format</span>
<span class="n">data</span> <span class="o">=</span> <span class="n">SnpArray</span><span class="p">(</span><span class="s">&quot;hapmap3&quot;</span><span class="p">)</span>

<span class="c"># Compute the PC of each SNP</span>
<span class="n">pcscore</span><span class="p">,</span><span class="n">_</span><span class="p">,</span><span class="n">_</span> <span class="o">=</span> <span class="n">pca</span><span class="p">(</span><span class="n">data</span><span class="p">)</span>
<span class="n">pcscore</span> <span class="o">=</span> <span class="n">convert</span><span class="p">(</span><span class="n">DataFrame</span><span class="p">,</span> <span class="n">pcscore</span><span class="p">)</span>
<span class="n">names!</span><span class="p">(</span><span class="n">pcscore</span><span class="p">,</span> <span class="p">[</span><span class="kt">Symbol</span><span class="p">(</span><span class="s">&quot;PC</span><span class="si">$i</span><span class="s">&quot;</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="kp">in</span> <span class="mi">1</span><span class="o">:</span><span class="n">size</span><span class="p">(</span><span class="n">pcscore</span><span class="p">,</span><span class="mi">2</span><span class="p">)])</span>

<span class="c"># Combine genotype data with PC results</span>
<span class="n">data</span> <span class="o">=</span> <span class="n">convert</span><span class="p">(</span><span class="n">DataFrame</span><span class="p">,</span> <span class="n">convert</span><span class="p">(</span><span class="kt">Matrix</span><span class="p">{</span><span class="kt">Float64</span><span class="p">},</span> <span class="n">data</span><span class="p">))</span>
<span class="n">data</span> <span class="o">=</span> <span class="n">hcat</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">pcscore</span><span class="p">)</span>

<span class="c"># Create the simulation model</span>
<span class="n">sim_model</span> <span class="o">=</span> <span class="n">FixedEffectModel</span><span class="p">(</span><span class="nd">@formula</span><span class="p">(</span><span class="n">T</span> <span class="o">~</span> <span class="n">x1</span><span class="o">+</span><span class="mi">2</span><span class="n">x2</span><span class="o">*</span><span class="n">x3</span><span class="o">+</span><span class="n">PC1</span><span class="o">+</span><span class="n">PC2</span><span class="p">),</span>
                             <span class="n">IdentityLink</span><span class="p">(),</span> <span class="n">NormalResponse</span><span class="p">(</span><span class="mf">1.0</span><span class="p">))</span>

<span class="c"># Generate the simulations</span>
<span class="n">y</span> <span class="o">=</span> <span class="n">simulate</span><span class="p">(</span><span class="n">sim_model</span><span class="p">,</span> <span class="n">data</span><span class="p">)</span>
</pre></div>


<p><a id='Simulate-a-trait-under-random-effect-model-1'></a></p>
<h3 id="simulate-a-trait-under-random-effect-model">Simulate a trait under random effect model</h3>
<p>The following code simulates a trait under the random effect model with two variance components, where one of the covariance matrix is the genetic relationship matrix (GRM) and the other the identity matrix. More specifically, the code below samples phenotype from</p>
<p>
<script type="math/tex; mode=display">
y \sim N(0, 0.8 K + 0.2I),
</script>
</p>
<p>where, <script type="math/tex">K</script> is the GRM estimated from genotype data using the <code>grm</code> function in <code>SnpArrays</code>.</p>
<div class="codehilite"><pre><span></span><span class="k">using</span> <span class="n">DataFrames</span><span class="p">,</span> <span class="n">Distributions</span><span class="p">,</span> <span class="n">SnpArrays</span><span class="p">,</span> <span class="n">TraitSimulation</span>

<span class="c"># Load Genotype data in PLINK format</span>
<span class="n">data</span> <span class="o">=</span> <span class="n">SnpArray</span><span class="p">(</span><span class="s">&quot;hapmap3&quot;</span><span class="p">)</span>
<span class="n">npeople</span> <span class="o">=</span> <span class="n">size</span><span class="p">(</span><span class="n">data</span><span class="p">,</span><span class="mi">1</span><span class="p">)</span>

<span class="c"># Compute GRM using the grm function in SnpArrays</span>
<span class="n">K</span> <span class="o">=</span> <span class="n">grm</span><span class="p">(</span><span class="n">data</span><span class="p">)</span>
<span class="nb">I</span> <span class="o">=</span> <span class="n">eye</span><span class="p">(</span><span class="n">npeople</span><span class="p">)</span>
<span class="n">Σ</span> <span class="o">=</span> <span class="p">[</span><span class="n">VarianceComponent</span><span class="p">(</span><span class="mf">0.8</span><span class="p">,</span> <span class="n">K</span><span class="p">),</span> <span class="n">VarianceComponent</span><span class="p">(</span><span class="mf">0.2</span><span class="p">,</span> <span class="nb">I</span><span class="p">)]</span>

<span class="c"># Create the simulation model (:T is the name of the simulated trait)</span>
<span class="n">sim_model</span> <span class="o">=</span> <span class="n">RandomEffectModel</span><span class="p">(</span><span class="o">:</span><span class="n">T</span><span class="p">,</span> <span class="n">Σ</span><span class="p">,</span> <span class="n">IdentityLink</span><span class="p">(),</span> <span class="n">NormalResponse</span><span class="p">(</span><span class="mf">1.0</span><span class="p">))</span>

<span class="c"># Generate the simulations</span>
<span class="n">y</span> <span class="o">=</span> <span class="n">simulate</span><span class="p">(</span><span class="n">sim_model</span><span class="p">,</span> <span class="n">data</span><span class="p">)</span>
</pre></div>


<p>We also provide the <code>@vc</code> macro to simplify the specification of the covariances. Note, the variables K and I must be defined before calling the <code>@vc</code> macro. For example, the following two lines of code are equivalent.</p>
<div class="codehilite"><pre><span></span><span class="n">Σ</span> <span class="o">=</span> <span class="nd">@vc</span> <span class="mf">0.8</span><span class="n">K</span> <span class="o">+</span> <span class="mf">0.2</span><span class="nb">I</span>
<span class="n">Σ</span> <span class="o">=</span> <span class="p">[</span><span class="n">VarianceComponent</span><span class="p">(</span><span class="mf">0.8</span><span class="p">,</span> <span class="n">K</span><span class="p">),</span> <span class="n">VarianceComponent</span><span class="p">(</span><span class="mf">0.2</span><span class="p">,</span> <span class="nb">I</span><span class="p">)]</span>
</pre></div>


<p><a id='Simulate-a-trait-under-mixed-effect-model-1'></a></p>
<h3 id="simulate-a-trait-under-mixed-effect-model">Simulate a trait under mixed effect model</h3>
<p>The following code simulates a trait under the mixed effect model with two variance components as well as a fixed effect <script type="math/tex">\mu</script>. More specifically, the code below samples phenotype from</p>
<p>
<script type="math/tex; mode=display">
y \sim N(\mu, 0.8 K + 0.2I),
</script>
</p>
<p>where, <script type="math/tex">\mu = x_1 + 2x_2x_3</script> the fixed effect vector <script type="math/tex">K</script> the GRM estimated from genotype data using the <code>grm</code> function in <code>SnpArrays</code>.</p>
<div class="codehilite"><pre><span></span><span class="k">using</span> <span class="n">DataFrames</span><span class="p">,</span> <span class="n">Distributions</span><span class="p">,</span> <span class="n">SnpArrays</span><span class="p">,</span> <span class="n">TraitSimulation</span>

<span class="c"># Load Genotype data in PLINK format</span>
<span class="n">data</span> <span class="o">=</span> <span class="n">SnpArray</span><span class="p">(</span><span class="s">&quot;hapmap3&quot;</span><span class="p">)</span>
<span class="n">npeople</span> <span class="o">=</span> <span class="n">size</span><span class="p">(</span><span class="n">data</span><span class="p">,</span><span class="mi">1</span><span class="p">)</span>

<span class="c"># Compute GRM using the grm function in SnpArrays</span>
<span class="n">K</span> <span class="o">=</span> <span class="n">grm</span><span class="p">(</span><span class="n">data</span><span class="p">)</span>
<span class="nb">I</span> <span class="o">=</span> <span class="n">eye</span><span class="p">(</span><span class="n">npeople</span><span class="p">)</span>
<span class="n">Σ</span> <span class="o">=</span> <span class="p">[</span><span class="n">VarianceComponent</span><span class="p">(</span><span class="mf">0.8</span><span class="p">,</span> <span class="n">K</span><span class="p">),</span> <span class="n">VarianceComponent</span><span class="p">(</span><span class="mf">0.2</span><span class="p">,</span> <span class="nb">I</span><span class="p">)]</span>
<span class="n">μ</span> <span class="o">=</span> <span class="nd">@formula</span><span class="p">(</span><span class="n">T</span> <span class="o">~</span> <span class="n">x1</span><span class="o">+</span><span class="mi">2</span><span class="n">x2</span><span class="o">*</span><span class="n">x3</span><span class="p">)</span>

<span class="c"># Create the simulation model (:T is the name of the simulated trait)</span>
<span class="n">sim_model</span> <span class="o">=</span> <span class="n">MixedEffectModel</span><span class="p">(</span><span class="n">μ</span><span class="p">,</span> <span class="n">Σ</span><span class="p">,</span> <span class="n">IdentityLink</span><span class="p">(),</span> <span class="n">NormalResponse</span><span class="p">(</span><span class="mf">1.0</span><span class="p">))</span>

<span class="c"># Generate the simulations</span>
<span class="n">y</span> <span class="o">=</span> <span class="n">simulate</span><span class="p">(</span><span class="n">sim_model</span><span class="p">,</span> <span class="n">data</span><span class="p">)</span>
</pre></div>


<p><a id='Simulate-a-trait-with-non-normal-distributions-1'></a></p>
<h3 id="simulate-a-trait-with-non-normal-distributions">Simulate a trait with non-normal distributions</h3>
<p>In the previous examples, all trait values are sampled from the normal distribution. However, users can also simulate traits with non-normal distributions, by specifying the type of distribution and the corresponding link function. For example, the following code simulate a trait under the Poisson distribution. More specifically, the following code samples phenotype from</p>
<p>
<script type="math/tex; mode=display">
y \sim Pois(\lambda)
</script>
</p>
<p>where</p>
<p>
<script type="math/tex; mode=display">
\lambda = \exp(x_1+2x_2x_3).
</script>
</p>
<div class="codehilite"><pre><span></span><span class="k">using</span> <span class="n">DataFrames</span><span class="p">,</span> <span class="n">Distributions</span><span class="p">,</span> <span class="n">SnpArrays</span><span class="p">,</span> <span class="n">TraitSimulation</span>

<span class="c"># Load genotype data in PLINK format</span>
<span class="n">data</span> <span class="o">=</span> <span class="n">SnpArray</span><span class="p">(</span><span class="s">&quot;hapmap3&quot;</span><span class="p">)</span>

<span class="c"># Create the fixed-effect simulation model</span>
<span class="n">sim_model</span> <span class="o">=</span> <span class="n">FixedEffectModel</span><span class="p">(</span><span class="nd">@formula</span><span class="p">(</span><span class="n">T</span> <span class="o">~</span> <span class="n">x1</span><span class="o">+</span><span class="mi">2</span><span class="n">x2</span><span class="o">*</span><span class="n">x3</span><span class="p">),</span>
                             <span class="n">LogLink</span><span class="p">(),</span> <span class="n">PoissonResponse</span><span class="p">())</span>

<span class="c"># Generate the simulation</span>
<span class="n">y</span> <span class="o">=</span> <span class="n">simulate</span><span class="p">(</span><span class="n">sim_model</span><span class="p">,</span> <span class="n">data</span><span class="p">)</span>
</pre></div>


<p><a id='Simulate-multiple-traits-1'></a></p>
<h2 id="simulate-multiple-traits">Simulate multiple traits</h2>
<p><a id='Simulate-three-independent-traits-with-different-response-distributions-1'></a></p>
<h3 id="simulate-three-independent-traits-with-different-response-distributions">Simulate three independent traits with different response distributions</h3>
<p>The following code simulates three independent traits (<script type="math/tex">y_1, y_2, y_3</script>) with different response distributions.</p>
<p>
<script type="math/tex; mode=display">
\mu_1 = 0.2x_1 + 3.0, y_1 \sim \text{Bin}(100, \mu_1)
</script>
</p>
<p>
<script type="math/tex; mode=display">
\mu_2 = 0.1x_3 + 2.0, y_2 \sim \text{Pois}(\mu_2)
</script>
</p>
<p>
<script type="math/tex; mode=display">
\mu_3 = 0.3x_4, y_3 \sim N(\mu_3, 2.0)
</script>
</p>
<div class="codehilite"><pre><span></span><span class="k">using</span> <span class="n">DataFrames</span><span class="p">,</span> <span class="n">Distributions</span><span class="p">,</span> <span class="n">SnpArrays</span><span class="p">,</span> <span class="n">TraitSimulation</span>

<span class="c"># Load genotype data in PLINK format</span>
<span class="n">data</span> <span class="o">=</span> <span class="n">SnpArray</span><span class="p">(</span><span class="s">&quot;hapmap3&quot;</span><span class="p">)</span>

<span class="c"># Create the simulation model</span>
<span class="n">μ</span> <span class="o">=</span> <span class="p">[</span><span class="n">y1</span> <span class="o">~</span> <span class="mf">0.2</span><span class="n">x1</span><span class="o">+</span><span class="mf">3.0</span><span class="p">,</span> <span class="n">y2</span> <span class="o">~</span> <span class="mf">0.1</span><span class="n">x3</span><span class="o">+</span><span class="mf">2.0</span><span class="p">,</span> <span class="n">y3</span> <span class="o">~</span> <span class="mf">0.3</span><span class="n">x4</span><span class="p">]</span>
<span class="n">link</span> <span class="o">=</span> <span class="p">[</span><span class="n">LogitLink</span><span class="p">(),</span> <span class="n">LogLink</span><span class="p">(),</span> <span class="n">IdentityLink</span><span class="p">()]</span>
<span class="n">dist</span> <span class="o">=</span> <span class="p">[</span><span class="n">BinomialResponse</span><span class="p">(</span><span class="mi">100</span><span class="p">),</span> <span class="n">PoissonResponse</span><span class="p">(),</span> <span class="n">NormalResponse</span><span class="p">(</span><span class="mf">2.0</span><span class="p">)]</span>
<span class="n">sim_model</span> <span class="o">=</span> <span class="n">FixedEffectModel</span><span class="p">(</span><span class="n">μ</span><span class="p">,</span> <span class="n">link</span><span class="p">,</span> <span class="n">dist</span><span class="p">)</span>

<span class="c"># Generate the simulations</span>
<span class="n">y</span> <span class="o">=</span> <span class="n">simulate</span><span class="p">(</span><span class="n">sim_model</span><span class="p">,</span> <span class="n">data</span><span class="p">)</span>
</pre></div>


<p><a id='Simulate-two-correlated-traits-1'></a></p>
<h3 id="simulate-two-correlated-traits">Simulate two correlated traits</h3>
<p>Users can also simulate two correlated traits. For example, the following code simulates two traits from</p>
<p>
<script type="math/tex; mode=display">
\begin{equation}
\left(
\begin{array}{c}
y_1 \\
y_2
\end{array}
\right)
\sim
N \left(
\left(            
\begin{array}{c}
x_1 + 0.2 \\
x_3 + 0.1
\end{array}
\right),
A \otimes K + B \otimes I       
\right)
\end{equation}
</script>
</p>
<p>where <script type="math/tex">A</script> and <script type="math/tex">B</script> are the cross covariances.</p>
<div class="codehilite"><pre><span></span><span class="k">using</span> <span class="n">DataFrames</span><span class="p">,</span> <span class="n">Distributions</span><span class="p">,</span> <span class="n">SnpArrays</span><span class="p">,</span> <span class="n">TraitSimulation</span>

<span class="c"># Load Genotype data in PLINK format</span>
<span class="n">data</span> <span class="o">=</span> <span class="n">SnpArray</span><span class="p">(</span><span class="s">&quot;hapmap3&quot;</span><span class="p">)</span>
<span class="n">npeople</span> <span class="o">=</span> <span class="n">size</span><span class="p">(</span><span class="n">data</span><span class="p">,</span><span class="mi">1</span><span class="p">)</span>

<span class="c"># Compute GRM using the grm function in SnpArrays</span>
<span class="n">K</span> <span class="o">=</span> <span class="n">grm</span><span class="p">(</span><span class="n">data</span><span class="p">)</span>
<span class="nb">I</span> <span class="o">=</span> <span class="n">eye</span><span class="p">(</span><span class="n">npeople</span><span class="p">)</span>
<span class="n">A</span> <span class="o">=</span> <span class="p">[</span><span class="mf">0.2</span> <span class="o">-</span><span class="mf">0.1</span><span class="p">;</span> <span class="o">-</span><span class="mf">0.1</span> <span class="mf">0.3</span><span class="p">]</span>
<span class="n">B</span> <span class="o">=</span> <span class="p">[</span><span class="mf">0.8</span> <span class="o">-</span><span class="mf">0.2</span><span class="p">;</span> <span class="o">-</span><span class="mf">0.2</span> <span class="mf">0.7</span><span class="p">]</span>
<span class="n">μ</span> <span class="o">=</span> <span class="p">[</span><span class="n">y1</span> <span class="o">~</span> <span class="n">x1</span><span class="o">+</span><span class="mf">0.2</span><span class="p">,</span> <span class="n">y2</span> <span class="o">~</span> <span class="n">x3</span><span class="o">+</span><span class="mf">0.1</span><span class="p">]</span>

<span class="c"># Create the simulation model</span>
<span class="n">sim_model</span> <span class="o">=</span> <span class="n">MixedEffectModel</span><span class="p">(</span><span class="n">μ</span><span class="p">,</span> <span class="p">(</span><span class="nd">@vc</span> <span class="n">A</span> <span class="n">⊗</span> <span class="n">K</span> <span class="o">+</span> <span class="n">B</span> <span class="n">⊗</span> <span class="nb">I</span><span class="p">),</span>
                             <span class="n">IdentityLink</span><span class="p">(),</span> <span class="n">NormalResponse</span><span class="p">(</span><span class="mf">1.0</span><span class="p">))</span>

<span class="c"># Generate the simulations</span>
<span class="n">y</span> <span class="o">=</span> <span class="n">simulate</span><span class="p">(</span><span class="n">sim_model</span><span class="p">,</span> <span class="n">data</span><span class="p">)</span>
</pre></div>
                
                  
                
              
              
                
              
            </article>
          </div>
        </div>
      </main>
      
        
<footer class="md-footer">
  
    <div class="md-footer-nav">
      <nav class="md-footer-nav__inner md-grid">
        
          <a href="../.." title="Introduction" class="md-flex md-footer-nav__link md-footer-nav__link--prev" rel="prev">
            <div class="md-flex__cell md-flex__cell--shrink">
              <i class="md-icon md-icon--arrow-back md-footer-nav__button"></i>
            </div>
            <div class="md-flex__cell md-flex__cell--stretch md-footer-nav__title">
              <span class="md-flex__ellipsis">
                <span class="md-footer-nav__direction">
                  Previous
                </span>
                Introduction
              </span>
            </div>
          </a>
        
        
          <a href="../../model/model/" title="Model Specifications" class="md-flex md-footer-nav__link md-footer-nav__link--next" rel="next">
            <div class="md-flex__cell md-flex__cell--stretch md-footer-nav__title">
              <span class="md-flex__ellipsis">
                <span class="md-footer-nav__direction">
                  Next
                </span>
                Model Specifications
              </span>
            </div>
            <div class="md-flex__cell md-flex__cell--shrink">
              <i class="md-icon md-icon--arrow-forward md-footer-nav__button"></i>
            </div>
          </a>
        
      </nav>
    </div>
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-footer-copyright">
        
        powered by
        <a href="http://www.mkdocs.org" title="MkDocs">MkDocs</a>
        and
        <a href="http://squidfunk.github.io/mkdocs-material/" title="Material for MkDocs">
          Material for MkDocs</a>
      </div>
      
        
      
    </div>
  </div>
</footer>
      
    </div>
    
      <script src="../../assets/javascripts/application-6940329b9f.js"></script>
      
      
      <script>app.initialize({url:{base:"../.."}})</script>
      
        <script src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS_HTML"></script>
      
        <script src="../../assets/mathjaxhelper.js"></script>
      
    
    
      
    
  </body>
</html>